src/libraries/System.Linq/src/System/Linq/Distinct.SpeedOpt.cs

```diff
@@ -17,7 +17,13 @@ private sealed partial class DistinctIterator<TSource>
 
             public override TSource? TryGetFirst(out bool found) => _source.TryGetFirst(out found);
 
-            public override bool Contains(TSource value) => _source.Contains(value);
+            public override bool Contains(TSource value) =>
+                // If we're using the default comparer, then source.Distinct().Contains(value) is no different from
+                // source.Contains(value), as the Distinct() won't remove anything that could have caused
+                // Contains to return true. If, however, there is a custom comparer, Distinct might remove
+                // the elements that would have matched, and thus we can't skip it.
+                _comparer is null || _comparer == EqualityComparer<TSource>.Default ? _source.Contains(value) :
```

Comment:
I'm guessing the reason we need to check both for `null` and `EqualityComparer<TSource>.Default` even though they functionally have the same behavior is because we use the latter representation for reference types?

Reply:
We don't _have_ to check for both; we could just check for null. But the latter is functionally equivalent and is substituted in for null in various places. I can remove it, though.

Reply:
Could the iterator constructor just normalize to one of the two representations?